### PR TITLE
Unify the block/grid strategy and implementation of ReduceLastDim and ReduceAny.

### PR DESCRIPTION
**Pull request by ZzSean**
### PR types

Performance optimization
### PR changes

OPs
### Describe

Unify the block/grid strategy and implementation of ReduceLastDim and ReduceAny
解决了在brocast反向时，reduce最后一维很小时性能差的问题，且优化了 ReduceLastDim的实现，提升了部分case性能。
| case                                |axis     | pytorch | 优化前  | 优化后  | 优化后相比pytorch | 加速比 |
| ---------------------------------------- | ------- | ------- | ----------------- | ------- | ----------------- | ------ |
| [4, 2048, 64, 128] | [2, 3]           | 305.15us | 328.35us | 303.54us | 打平 (0.53%)      | 1.08   |
| [16, 2048, 7, 7] | [2, 3]             | 21.738us | 49.712us |  16.133us | 优于 (34.74%)      | 3.08   |
| [512, 896, 4, 12] | [3]      | 214.61us | 2.5947ms | 166.01us | 优于 (29.28%)      | 15.63   |

### CODE REVIEW

**Comment by AnnaTrainingG** [comment]
LGTM

File: paddle/fluid/operators/reduce_ops/reduce_op.cu.h [hunk 1]
```diff
@@ -524,18 +531,20 @@ static __device__ T WarpReduce(T val, ReduceOp reducer) {
 template <typename T, typename ReduceOp>
 static __device__ T BlockXReduce(T val, ReduceOp reducer) {
   using detail::kWarpSize;
-  __shared__ T shared[kWarpSize];
+  __shared__ T shared[2 * kWarpSize];
```
Comment (Xreki) [inline paddle/fluid/operators/reduce_ops/reduce_op.cu.h:534]: 为什么是`2 * kWarpSize`？后续可以考虑继续优化下这个函数，使得它可以作为一个基础函数，可以适用于如softmax、batch_norm等算子的使用场景。

**Review by Xreki — APPROVED** [review]
LGTM
